What to do if client Ping fail?

I have a kafka consumer that perform a Ping with a timeout of 5 seconds

if it fails I am calling Close() and return the error because... it seems the right thing to do

it works fine if the domain does not exist (network issue) or with a wrong password (auth issue), but I find one recent issue when we omit any SASL by mistake and the code took around one minute to finish.

Example of error

```
broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?
```

first I check if Ping was ignoring the context timeout, but it hangs on Close()

even if I call LeaveGroupContext before, it still take a lot of time

the only scenario where it is quicker is if I use a context with a very short timeout on LeaveGroupContext

based on this, what should I do if the Ping returns an error?
 
it is fine just leave the client open? 
or there is some retry configuration that may helps?
perhaps should I call client.Close in a goroutine to be eventually finish?

I did not try kfake,

I tested using redpanda (via testcontainers) with SASL scram 256. I saw this issue in production when trying to connect to kafka with sasl plain (but I had to use redpanda in my tests because the setup was easier, and the result is the same. 

for instance, with mu workaound (call LeaveGroupContext after ping and before Close)  it generates some log lines like this

```console
2025/04/30 10:32:15 INFO immediate metadata update triggered component=kafka kind=consumer why="querying metadata for consumer initialization"
2025/04/30 10:32:15 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=ApiVersions addr=localhost:35319 broker=seed_0 err=EOF
2025/04/30 10:32:15 INFO assigning partitions component=kafka kind=consumer why="invalidating all assignments in LeaveGroup" how="unassigning everything" input=""
2025/04/30 10:32:15 INFO leaving group component=kafka kind=consumer group=foo-bar-28b135e8-912f-549c-b1c9-fd32476d7af3 member_id=""
2025/04/30 10:32:15 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35319 broker=seed_0 err=EOF
2025/04/30 10:32:15 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35319 broker=seed_0 err=EOF
2025/04/30 10:32:15 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35319 broker=seed_0 err=EOF
2025/04/30 10:32:16 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35319 broker=seed_0 err=EOF
2025/04/30 10:32:16 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35319 broker=seed_0 err=EOF
2025/04/30 10:32:16 INFO assigning partitions component=kafka kind=consumer why="invalidating all assignments in LeaveGroup" how="unassigning everything" input=""
```

now, if I only call Close after Ping, it tooks ~1 minute and we can see the same log lines over and over, like it is trying to update something during the closing process.

```console
=== RUN   TestIntegrationConsumerWithAuthentication
=== PAUSE TestIntegrationConsumerWithAuthentication
=== CONT  TestIntegrationConsumerWithAuthentication
2025/04/30 10:23:54 github.com/testcontainers/testcontainers-go - Connected to docker: 
  Server Version: 28.1.1
  API Version: 1.49
  Operating System: Ubuntu 24.04.2 LTS
  Total Memory: 15720 MB
  Testcontainers for Go Version: v0.37.0
  Resolved Docker Host: unix:///var/run/docker.sock
  Resolved Docker Socket Path: /var/run/docker.sock
  Test SessionID: deb69685d191514d388f98df3813d4c603b23c4b0d5c9b6408492b9b0bbd6953
  Test ProcessID: 9f029347-8cb9-438b-b8d0-bb35aa5afd61
    lifecycle.go:79: 🐳 Creating container for image docker.redpanda.com/redpandadata/redpanda:v23.3.3
    lifecycle.go:79: 🐳 Creating container for image testcontainers/ryuk:0.11.0
    lifecycle.go:85: ✅ Container created: 05880c43cd25
    lifecycle.go:91: 🐳 Starting container: 05880c43cd25
    lifecycle.go:97: ✅ Container started: 05880c43cd25
    lifecycle.go:287: ⏳ Waiting for container id 05880c43cd25 image: testcontainers/ryuk:0.11.0. Waiting for: &{Port:8080/tcp timeout:<nil> PollInterval:100ms skipInternalCheck:false}
    lifecycle.go:103: 🔔 Container is ready: 05880c43cd25
    lifecycle.go:85: ✅ Container created: 3b7be1772362
    lifecycle.go:91: 🐳 Starting container: 3b7be1772362
    lifecycle.go:97: ✅ Container started: 3b7be1772362
    lifecycle.go:287: ⏳ Waiting for container id 3b7be1772362 image: docker.redpanda.com/redpandadata/redpanda:v23.3.3. Waiting for: &{timeout:<nil> deadline:<nil> Strategies:[0xc00030dcb0 0xc00030dce0 0xc00030dd10]}
    lifecycle.go:103: 🔔 Container is ready: 3b7be1772362
=== RUN   TestIntegrationConsumerWithAuthentication/without_authentication
=== PAUSE TestIntegrationConsumerWithAuthentication/without_authentication
=== CONT  TestIntegrationConsumerWithAuthentication/without_authentication
2025/04/30 10:23:56 INFO just push records to kafka topic=foo-28b135e8-912f-549c-b1c9-fd32476d7af3 n=0
2025/04/30 10:23:56 INFO KafkaConsumer/Consume start
2025/04/30 10:23:56 INFO immediate metadata update triggered component=kafka kind=consumer why="querying metadata for consumer initialization"
2025/04/30 10:23:56 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=ApiVersions addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:23:56 INFO assigning partitions component=kafka kind=consumer why="invalidating all assignments in LeaveGroup" how="unassigning everything" input=""
2025/04/30 10:23:56 INFO leaving group component=kafka kind=consumer group=foo-bar-28b135e8-912f-549c-b1c9-fd32476d7af3 member_id=""
2025/04/30 10:23:56 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:23:56 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:23:56 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:23:57 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:23:57 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:23:57 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:23:57 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:23:57 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:23:58 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:23:58 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:23:58 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:23:58 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:23:59 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:23:59 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:00 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:00 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:00 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:01 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:01 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:02 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:03 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:03 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:03 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:03 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:05 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:06 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:06 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:06 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:07 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:07 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:09 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:09 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:09 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:10 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:10 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:12 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:12 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:12 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:13 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:13 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:15 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:16 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:16 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:16 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:16 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:17 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:19 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:19 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:19 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:20 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:20 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:22 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:22 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:22 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:22 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:23 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:25 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:25 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:25 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:25 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:25 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:26 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:26 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:26 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:27 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:29 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:29 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:29 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:29 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:29 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:31 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:32 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:32 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:32 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:33 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:34 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:35 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:35 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:35 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:36 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:36 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:38 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:38 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:39 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:39 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:39 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:41 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:42 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:42 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:42 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:42 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:44 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:45 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:45 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:45 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:46 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:46 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:48 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:48 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:48 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:49 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:49 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:51 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:51 INFO metadata update triggered component=kafka kind=consumer why="re-updating metadata due to err: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?"
2025/04/30 10:24:51 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:52 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:52 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=Metadata addr=localhost:35306 broker=seed_0 err=EOF
2025/04/30 10:24:54 WARN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=kafka kind=consumer req=FindCoordinator addr=localhost:35306 broker=seed_0 err=EOF
    consumer_kafka_integration_test.go:396: 
        	Error Trace:	/home/tiago/work/go/src/gitlab.nereides.weborama.com/go/lib-kafka/consumer/consumer_kafka_integration_test.go:396
        	Error:      	"11s" is not greater than "57.726598072s"
        	Test:       	TestIntegrationConsumerWithAuthentication/without_authentication
        	Messages:   	expects wait max 11s, got 57.726598072s
=== NAME  TestIntegrationConsumerWithAuthentication
    lifecycle.go:109: 🐳 Stopping container: 3b7be1772362
    lifecycle.go:115: ✅ Container stopped: 3b7be1772362
    lifecycle.go:121: 🐳 Terminating container: 3b7be1772362
    lifecycle.go:127: 🚫 Container terminated: 3b7be1772362
--- FAIL: TestIntegrationConsumerWithAuthentication (2.18s)

```

--- a/pkg/kgo/client.go
+++ b/pkg/kgo/client.go
@@ -1804,7 +1804,8 @@ func (cl *Client) doLoadCoordinators(ctx context.Context, typ int8, keys ...stri
 			"coordinator_keys", req.CoordinatorKeys,
 		)
 
-		shards := cl.RequestSharded(cl.ctx, req)
+		ctx := context.WithValue(cl.ctx, noShardRetryCtx, true)
+		shards := cl.RequestSharded(ctx, req)
 
 		for _, shard := range shards {
 			if shard.Err != nil {
@@ -2316,6 +2317,8 @@ type sharder interface {
 	merge([]ResponseShard) (kmsg.Response, error)
 }
 
+var noShardRetryCtx = func() *string { s := "no_shard_retry"; return &s }()
+
 // handleShardedReq splits and issues requests to brokers, recursively
 // splitting as necessary if requests fail and need remapping.
 func (cl *Client) handleShardedReq(ctx context.Context, req kmsg.Request) ([]ResponseShard, shardMerge) {
@@ -2388,6 +2391,7 @@ func (cl *Client) handleShardedReq(ctx context.Context, req kmsg.Request) ([]Res
 
 	l := cl.cfg.logger
 	debug := l.Level() >= LogLevelDebug
+	noRetries := ctx != nil && ctx.Value(noShardRetryCtx) != nil
 
 	// issue is called to progressively split and issue requests.
 	//
@@ -2479,7 +2483,7 @@ func (cl *Client) handleShardedReq(ctx context.Context, req kmsg.Request) ([]Res
 				backoff := cl.cfg.retryBackoff(tries)
 				if err != nil &&
 					(reshardable && isPinned && errors.Is(err, errBrokerTooOld) && tries <= 3) ||
-					(retryTimeout == 0 || time.Now().Add(backoff).Sub(start) <= retryTimeout) && cl.shouldRetry(tries, err) && cl.waitTries(ctx, backoff) {
+					(retryTimeout == 0 || time.Now().Add(backoff).Sub(start) <= retryTimeout) && cl.shouldRetry(tries, err) && cl.waitTries(ctx, backoff) && !noRetries {
 					// Non-reshardable re-requests just jump back to the
 					// top where the broker is loaded. This is the case on
 					// requests where the original request is split to

--- a/pkg/kgo/client.go
+++ b/pkg/kgo/client.go
@@ -1346,11 +1346,11 @@ func (cl *Client) retryableBrokerFn(fn func() (*broker, error)) *retryable {
 }
 
 func (cl *Client) shouldRetry(tries int, err error) bool {
-	return (kerr.IsRetriable(err) || isRetryableBrokerErr(err)) && int64(tries) < cl.cfg.retries
+	return (kerr.IsRetriable(err) || isRetryableBrokerErr(err)) && int64(tries) <= cl.cfg.retries
 }
 
 func (cl *Client) shouldRetryNext(tries int, err error) bool {
-	return isSkippableBrokerErr(err) && int64(tries) < cl.cfg.retries
+	return isSkippableBrokerErr(err) && int64(tries) <= cl.cfg.retries
 }
 
 type retryable struct {
@@ -1406,7 +1406,7 @@ start:
 	}
 
 	if err != nil || retryErr != nil {
-		if r.limitRetries == 0 || tries < r.limitRetries {
+		if r.limitRetries == 0 || tries <= r.limitRetries {
 			backoff := r.cl.cfg.retryBackoff(tries)
 			if retryTimeout == 0 || time.Now().Add(backoff).Sub(tryStart) <= retryTimeout {
 				// If this broker / request had a retryable error, we can
@@ -1415,8 +1415,10 @@ start:
 				// broker is different than the current, we also retry.
 				if r.cl.shouldRetry(tries, err) || r.cl.shouldRetry(tries, retryErr) {
 					r.cl.cfg.logger.Log(LogLevelDebug, "retrying request",
+						"request", kmsg.NameForKey(req.Key()),
 						"tries", tries,
 						"backoff", backoff,
+						"time_since_start", time.Since(tryStart),
 						"request_error", err,
 						"response_error", retryErr,
 					)

--- a/pkg/kgo/client.go
+++ b/pkg/kgo/client.go
@@ -2485,10 +2485,10 @@ func (cl *Client) handleShardedReq(ctx context.Context, req kmsg.Request) ([]Res
 					// requests where the original request is split to
 					// dedicated brokers; we do not want to re-shard that.
 					if !reshardable {
-						l.Log(LogLevelDebug, "sharded request failed, reissuing without resharding", "req", kmsg.Key(myIssue.req.Key()).Name(), "time_since_start", time.Since(start), "tries", try.tries, "err", err)
+						l.Log(LogLevelDebug, "sharded request failed, reissuing without resharding", "req", kmsg.Key(myIssue.req.Key()).Name(), "time_since_start", time.Since(start), "tries", tries, "err", err)
 						goto start
 					}
-					l.Log(LogLevelDebug, "sharded request failed, resharding and reissuing", "req", kmsg.Key(myIssue.req.Key()).Name(), "time_since_start", time.Since(start), "tries", try.tries, "err", err)
+					l.Log(LogLevelDebug, "sharded request failed, resharding and reissuing", "req", kmsg.Key(myIssue.req.Key()).Name(), "time_since_start", time.Since(start), "tries", tries, "err", err)
 					issue(reqTry{tries, myUnderlyingReq, err})
 					return
 				}

--- a/pkg/kgo/config.go
+++ b/pkg/kgo/config.go
@@ -507,7 +507,7 @@ func defaultCfg() cfg {
 			return func(fails int) time.Duration {
 				const (
 					min = 250 * time.Millisecond
-					max = 5 * time.Second / 2
+					max = 5 * time.Second
 				)
 				if fails <= 0 {
 					return min

--- a/pkg/kgo/consumer_group.go
+++ b/pkg/kgo/consumer_group.go
@@ -2333,9 +2333,7 @@ func (g *groupConsumer) getUncommittedLocked(head, dirty bool) map[string]map[in
 	return uncommitted
 }
 
-type commitContextFnT struct{}
-
-var commitContextFn commitContextFnT
+var commitContextFn = func() *string { s := "commit_ctx"; return &s }()
 
 // PreCommitFnContext attaches fn to the context through WithValue. Using the
 // context while committing allows fn to be called just before the commit is
@@ -2346,9 +2344,7 @@ func PreCommitFnContext(ctx context.Context, fn func(*kmsg.OffsetCommitRequest) 
 	return context.WithValue(ctx, commitContextFn, fn)
 }
 
-type txnCommitContextFnT struct{}
-
-var txnCommitContextFn txnCommitContextFnT
+var txnCommitContextFn = func() *string { s := "txn_commit_ctx"; return &s }()
 
 // PreTxnCommitFnContext attaches fn to the context through WithValue. Using
 // the context while committing a transaction allows fn to be called just

--- a/pkg/kgo/errors.go
+++ b/pkg/kgo/errors.go
@@ -63,6 +63,12 @@ func isRetryableBrokerErr(err error) bool {
 	// EOF can be returned if a broker kills a connection unexpectedly, and
 	// we can retry that. Same for ErrClosed.
 	if errors.Is(err, net.ErrClosed) || errors.Is(err, io.EOF) {
+		// If the FIRST read is EOF, that is usually not a good sign,
+		// often it's from bad SASL. We err on the side of pessimism
+		// and do not retry.
+		if ee := (*ErrFirstReadEOF)(nil); errors.As(err, &ee) {
+			return false
+		}
 		return true
 	}
 	// We could have a retryable producer ID failure, which then bubbled up

--- a/pkg/kgo/metadata.go
+++ b/pkg/kgo/metadata.go
@@ -239,6 +239,7 @@ loop:
 		}
 
 		retryWhy, err := cl.updateMetadata()
+		lastAt = time.Now()
 		if retryWhy != nil || err != nil {
 			// If err is non-nil, the metadata request failed
 			// itself and already retried 3x; we do not loop more.
@@ -281,12 +282,13 @@ loop:
 		if err == nil {
 			cl.metawait.signal()
 			cl.consumer.doOnMetadataUpdate()
-			lastAt = time.Now()
 			consecutiveErrors = 0
 			continue
 		}
 
 		consecutiveErrors++
+		// We sleep a bit in case the max metadata age is very small;
+		// typically this sleep is inconsequential.
 		after := time.NewTimer(cl.cfg.retryBackoff(consecutiveErrors))
 	backoff:
 		select {

